lib/rpmfi.c

```diff
@@ -1508,53 +1512,70 @@ static void rpmfilesBuildNLink(rpmfiles fi, Header h)
     return;
 }
 
-/* Convert a tag of hex strings to binary presentation */
-/* If lengths is non-NULL, assume variable length strings */
-static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len,
-			int **lengths, int *maxlen)
+/*
+ * Convert a tag of variable len hex strings to binary presentation,
+ * accessed via offsets to a contiguous binary blob. Empty values
+ * are represented by identical consequtive offsets. The offsets array
+ * always has one extra element to allow calculating the size of the
+ * last element.
+ */
+static uint8_t *hex2binv(Header h, rpmTagVal tag, rpm_count_t num,
+			uint32_t **offsetp)
 {
     struct rpmtd_s td;
     uint8_t *bin = NULL;
+    uint32_t *offs = NULL;
 
     if (headerGet(h, tag, &td, HEADERGET_MINMEM) && rpmtdCount(&td) == num) {
 	const char *s;
-	int maxl = 0;
-	int *lens = NULL;
-
-	/* Figure string sizes + max length for allocation purposes */
-	if (lengths) {
-	    int i = 0;
-	    lens = xmalloc(num * sizeof(*lens));
-
-	    while ((s = rpmtdNextString(&td))) {
-		lens[i] = strlen(s) / 2;
-		if (lens[i] > maxl)
-		    maxl = lens[i];
-		i++;
+	int i = 0;
+	uint8_t *t = bin = xmalloc(((rpmtdSize(&td) / 2) + 1));
+	offs = xmalloc((num + 1) * sizeof(*offs));
+
+	while ((s = rpmtdNextString(&td))) {
+	    uint32_t slen = strlen(s);
+	    uint32_t len = slen / 2;
+	    if (slen % 2) {
+		bin = rfree(bin);
+		offs = rfree(offs);
+		goto exit;
 	    }
+	    offs[i] = t - bin;
+	    for (int j = 0; j < len; j++, t++, s += 2)
+		*t = (rnibble(s[0]) << 4) | rnibble(s[1]);
+	    i++;
+	}
+	offs[i] = t - bin;
+	*offsetp = offs;
+    }
 
-	    *lengths = lens;
-	    *maxlen = maxl;
+exit:
+    rpmtdFreeData(&td);
+    return bin;
+}
 
-	    /* Reinitialize iterator for next round */
-	    rpmtdInit(&td);
-	} else {
-	    maxl = len;
-	}
+/* Convert a tag of hex strings to binary presentation */
+static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len)
+{
+    struct rpmtd_s td;
+    uint8_t *bin = NULL;
+
+    if (headerGet(h, tag, &td, HEADERGET_MINMEM) && rpmtdCount(&td) == num) {
+	uint8_t *t = bin = xmalloc(num * len);
+	const char *s;
 
-	uint8_t *t = bin = xmalloc(num * maxl);
-	int i = 0;
 	while ((s = rpmtdNextString(&td))) {
+	    if (strlen(s) % 2) {
+		bin = rfree(bin);
+		break;
```

Comment:
This only checks that the length is even, not that it is correct.

Reply:
Erm. These strings are variable length by definition, the only thing we *can* check is that it's even and thus wont overrun.

Reply:
Oh, that was in the hex2bin() context. GH review UI is such a useless piece of crap.
Yeah in that case we could check the length, but we're well into fixing unrelated stuff here now.

Reply:
Done.